### PR TITLE
EventGrid: Powershell work for 2018-09-15-preview

### PR DESCRIPTION
**Pull request by ahamad-MS**
## Description

* Added new cmdlets to manage the resources in Azure Event Grid service.

    - New-AzureRmEventGridDomain
        - Creates a new Azure Event Grid Domain.
    - Get-AzureRmEventGridDomain
        - Gets the details of an Event Grid Domain, or gets a list of all Event Grid Domains in the current Azure subscription.
    - Remove-AzureRmEventGridDomain
        - Removes an Azure Event Grid Domain.
    - New-AzureRmEventGridTDomainKey
        - Regenerates the shared access key for an Azure Event Grid Domain.
    - Get-AzureRmEventGridDomainKey
        - Gets the shared access keys used to publish events to an Event Grid Domain.
    - Get-AzureRmEventGridDomainTopic
        - Gets the details of an Event Grid Domain Topic, or gets a list of all Event Grid Domain Topics under specific Event Grid Domain in the current Azure subscription.

* Update the following cmdlets to support new scenario in 2018-09-15-preview API version
    - New-AzureRmEventGridSubscription: 
        - Add new mandatory parameters to support piping for the new Event Grid Domain and Event Grid Domain Topic to allow creating new event subscription under these resources.
        - Add new mandatory parameters for specifying the new Event Grid Domain name and/or Event Grid Domain Topic name to allow creating new event subscription under these resources.
        - Add new Parameter sets for domains and domain topics to allow reusing existing parameters (e.g., EndPointType, SubjectBeginsWith, etc).
        - Add new optional parameters for specifying: 
            - Event subscription expiration date, 
            - Advanced filtering parameters. 
    - Update-AzureRmEventGridSubscription: Add new optional parameters for specifying: 
        - Similar as New-AzureRmEventGridSubscription.
    - Remove-AzureRmEventGridSubscription
        - Add new mandatory parameters to support piping for Event Grid Domain and Event Grid Domain Topic to allow removing existing event subscription under these resources.
        - Add new mandatory parameters for specifying the Event Grid Domain name and/or Event Grid Domain Topic name to allow removing existing event subscription under these resources.

Detailed description of the work is available here https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/164

## Checklist

- [X] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [X] The title of the PR is clear and informative
- [X] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [X] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [X] If applicable, the changes made in the PR have proper test coverage
- [X] For public API changes to cmdlets:
    - [X] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (Microsoft internal partners only)
    - [X] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by adxsdkps** [comment]
Can one of the admins verify this patch?

**Comment by ahamad-MS** [comment]
@cormacpayne Hello Cormac. It feels like that continuous-integration/travis-ci and default checks are stuck for the last 2 hours in "Expected --- Waiting for status to be reported" . I don't see a link where I can check the progress in these checks or what is the expected status it is waiting for. I tried to push new fix for the md files hoping it will change this status but no luck.  Any idea?

**Comment by cormacpayne** [comment]
@azuresdkci add to whitelist

**Comment by ahamad-MS** [comment]
@cormacpayne 
Can you please advise on how to resolve the CI issue? It is not recognizing the usage of the private Microsoft.Azure.Management.EventGrid nuget submitted in tools\localfeed. 

Domain/GetAzureEventGridDomain.cs(19,34): error CS0234: The type or namespace name 'EventGrid' does not exist in the namespace 'Microsoft.Azure.Management' (are you missing an assembly reference?) [/home/travis/build/Azure/azure-powershell/src/ResourceManager/EventGrid/Commands.EventGrid/Commands.EventGrid.Netcore.csproj] [/home/travis/build/Azure/azure-powershell/build.proj]

Domain/GetAzureEventGridDomainKeys.cs(18,34): error CS0234: The type or namespace name 'EventGrid' does not exist in the namespace 'Microsoft.Azure.Management' (are you missing an assembly reference?) [/home/travis/build/Azure/azure-powershell/src/ResourceManager/EventGrid/Commands.EventGrid/Commands.EventGrid.Netcore.csproj] [/home/travis/build/Azure/azure-powershell/build.proj]

File: src/ResourceManager/EventGrid/Commands.EventGrid/Commands.EventGrid.Netcore.csproj [hunk 1]
```diff
@@ -28,7 +28,10 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.Azure.Management.EventGrid" Version="2.0.0-preview" />
+    <Reference Include="Microsoft.Azure.Management.EventGrid">
```
Comment (cormacpayne) [inline src/ResourceManager/EventGrid/Commands.EventGrid/Commands.EventGrid.Netcore.csproj]: @ahamad-MS revert the change made in this file and then just change `2.0.0-preview` to `3.0.0-preview`

**Comment by maddieclayton** [comment]
@ahamad-MS In order to publish this module, you will also need to rerelease your SDK management client with a new Newtonsoft Json dependency of 10.0.3 (we just completed the work of upgrading the stable PowerShell module to Newtonsoft Json 10 today). As we have already upgraded the SDK repo to Newtonsoft Json 10, all you will need to do is open a PR to bump the version of your SDK client from 3.0.0-preview to 3.0.1-preview, and publish from the Azure/psSdkJson6 branch once your PR is merged.  Then, when you pull latest from Azure/azure-powershell preview branch, your module should be good to go.

File: src/ResourceManager/EventGrid/Commands.EventGrid.Test/Commands.EventGrid.Test.csproj [hunk 2]
```diff
@@ -40,15 +40,31 @@
     <Prefer32Bit>false</Prefer32Bit>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="Microsoft.Azure.Management.EventGrid, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+    <Reference Include="Hyak.Common, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (cormacpayne) [inline src/ResourceManager/EventGrid/Commands.EventGrid.Test/Commands.EventGrid.Test.csproj]: @ahamad-MS there's a few package references being added here that are already being referenced from a separate file (`Common.Dependencies.Test.targets`), so you can remove these and ensure the only package reference added/changed was `Microsoft.Azure.Management.EventGrid`

File: src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomain.cs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        VerbsCommon.Get,
+        ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "EventGridDomain",
+        DefaultParameterSetName = ResourceGroupNameParameterSet),
+    OutputType(typeof(PSDomain), typeof(PSDomainListInstance))]
+
+    public class GetAzureEventGridDomain : AzureEventGridCmdletBase
+    {
+        [Parameter(
+            Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
```
Comment (cormacpayne) [inline src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomain.cs]: @ahamad-MS nit: the `-ResourceId` parameter should be the only one in this cmdlet that has `ValueFromPipelineByPropertyName = true`

File: src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomainKeys.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    [Cmdlet(
+        VerbsCommon.Get,
+        ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "EventGridDomainKey",
+        DefaultParameterSetName = DomainNameParameterSet),
+    OutputType(typeof(DomainSharedAccessKeys))]
+
+    public class GetAzureEventGridDomainKeys : AzureEventGridCmdletBase
+    {
+        [Parameter(Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
```
Comment (cormacpayne) [inline src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomainKeys.cs]: @ahamad-MS nit: same comment about removing the `ValueFromPipelineByPropertyName` property from everything by `-ResourceId`

File: src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomainKeys.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        [Alias("DomainName")]
+        public string Name { get; set; }
+
+        [Parameter(Mandatory = true,
+            ValueFromPipeline = true,
+            Position = 0,
+            HelpMessage = EventGridConstants.DomainInputObjectHelp,
+            ParameterSetName = DomainInputObjectParameterSet)]
+        [ValidateNotNullOrEmpty]
+        public PSDomain InputObject { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomainKeys.cs]: @ahamad-MS nit: I would rename this to `DomainObject` -- usually the `InputObject` parameter is used to represent the noun of the cmdlet (which in this case is the domain key)

File: src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomainKeys.cs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+                resourceGroupName = this.InputObject.ResourceGroupName;
+                domainName = this.InputObject.DomainName;
+            }
+            else
+            {
+                resourceGroupName = this.ResourceGroupName;
+                domainName = this.Name;
+            }
+
+            this.WriteObject(this.Client.GetDomainSharedAccessKeys(resourceGroupName, domainName));
```
Comment (cormacpayne) [inline src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomainKeys.cs]: @ahamad-MS does this return a list? If so, then it should be `WriteObject(this.Client.GetDomainSharedAccessKeys(resourceGroupName, domainName), true);`

File: src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomainTopic.cs [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        VerbsCommon.Get,
+        ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "EventGridDomainTopic",
+        DefaultParameterSetName = DomainTopicNameParameterSet),
+    OutputType(typeof(PSDomainTopic), typeof(PSDomainTopicListInstance))]
+
+    public class GetAzureEventGridDomainTopic : AzureEventGridCmdletBase
+    {
+        [Parameter(
+            Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
```
Comment (cormacpayne) [inline src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomainTopic.cs]: @ahamad-MS nit: same comment about `ValueFromPipelineByPropertyName`

File: src/ResourceManager/EventGrid/Commands.EventGrid/Domain/NewAzureEventGridDomain.cs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        [LocationCompleter("Microsoft.EventGrid/domains")]
+        [ValidateNotNullOrEmpty]
+        public string Location { get; set; }
+
+        /// <summary>
+        /// Hashtable which represents resource Tags.
+        /// </summary>
+        [Parameter(
+            Mandatory = false,
+            Position = 3,
```
Comment (cormacpayne) [inline src/ResourceManager/EventGrid/Commands.EventGrid/Domain/NewAzureEventGridDomain.cs]: @ahamad-MS optional parameters should not have a position, so you can remove the position from this parameter

File: src/ResourceManager/EventGrid/Commands.EventGrid/Domain/NewAzureEventGridDomainKey.cs [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "EventGridDomainKey",
+        DefaultParameterSetName = DomainNameParameterSet,
+        SupportsShouldProcess = true), 
+    OutputType(typeof(DomainSharedAccessKeys))]
+
+    public class NewAzureEventGridDomainKey : AzureEventGridCmdletBase
+    {
+        [Parameter(
+            Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
```
Comment (cormacpayne) [inline src/ResourceManager/EventGrid/Commands.EventGrid/Domain/NewAzureEventGridDomainKey.cs]: @ahamad-MS nit: same comment about removing `ValueFromPipelineByPropertyName`

File: src/ResourceManager/EventGrid/Commands.EventGrid/Domain/RemoveAzureEventGridDomain.cs [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        VerbsCommon.Remove,
+        ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "EventGridDomain",
+        DefaultParameterSetName = DomainNameParameterSet,
+        SupportsShouldProcess = true),
+    OutputType(typeof(bool))]
+
+    public class RemoveAzureEventGridDomain : AzureEventGridCmdletBase
+    {
+        [Parameter(Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
```
Comment (cormacpayne) [inline src/ResourceManager/EventGrid/Commands.EventGrid/Domain/RemoveAzureEventGridDomain.cs]: @ahamad-MS nit: same comment about removing `ValueFromPipelineByPropertyName` from all cmdlets except `-ResourceId`

File: src/ResourceManager/Sql/Commands.Sql/help/New-AzureRmSqlDatabase.md [hunk 11]
```diff
@@ -1,3 +1,478 @@
+---
```
Comment (cormacpayne) [inline src/ResourceManager/Sql/Commands.Sql/help/New-AzureRmSqlDatabase.md]: @ahamad-MS please revert the changes made to this file and the `AzureRM.psm1` file below this one

File: src/ResourceManager/EventGrid/Commands.EventGrid.Test/Commands.EventGrid.Test.Netcore.csproj [hunk 12]
```diff
@@ -24,12 +24,9 @@
     <DefineConstants>TRACE;RELEASE;NETSTANDARD;SIGN</DefineConstants>
   </PropertyGroup>
 
-  <ItemGroup Condition="'$(TargetFramework)' != 'net472'">
```
[the remark this answers is not in the document]
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid.Test/Commands.EventGrid.Test.Netcore.csproj]: Fixed

File: src/ResourceManager/EventGrid/Commands.EventGrid.Test/Commands.EventGrid.Test.csproj (hunk 2 above)
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid.Test/Commands.EventGrid.Test.csproj]: Fixed

File: src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomain.cs (hunk 3 above)
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomain.cs]: Fixed everywhere

File: src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomainKeys.cs (hunk 5 above)
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomainKeys.cs]: Fixed here for DomainObject and for  TopicObject in GetAzureEventGridTopicKeys.cs

File: src/ResourceManager/EventGrid/Commands.EventGrid/Domain/NewAzureEventGridDomain.cs (hunk 8 above)
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/Domain/NewAzureEventGridDomain.cs]: Fixed here and everywhere in the code

**Comment by ahamad-MS** [comment]
@maddieclayton  New PR for the SDK 3.0.1 is avaiable here: https://github.com/Azure/azure-sdk-for-net/pull/4980

File: src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomain.cs (hunk 3 above)
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomain.cs]: @cormacpayne  Taking this back. This is actually needed here to support the following scenario

Get-AzureRmEventGridDomain -ResourceGroupName aegBugbashResourceGroup2 -Name aegBugBashDomain2 | Get-AzureRmEventGridDomainKey

Otherwise, we will get his error 
Get-AzureRmEventGridDomainKey : The input object cannot be bound to any parameters for the command either because the command does not take pipeline input or the input and its
properties do not match any of the parameters that take pipeline input

File: src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomainKeys.cs (hunk 4 above)
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomainKeys.cs]: @cormacpayne  This is actually needed here to support the following scenario
Get-AzureRmEventGridDomain -ResourceGroupName aegBugbashResourceGroup2 -Name aegBugBashDomain2 | Get-AzureRmEventGridDomainKey
Otherwise, we will get his error
Get-AzureRmEventGridDomainKey : The input object cannot be bound to any parameters for the command either because the command does not take pipeline input or the input and its
properties do not match any of the parameters that take pipeline input

File: src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomainTopic.cs (hunk 7 above)
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/Domain/GetAzureEventGridDomainTopic.cs]: @cormacpayne Same here. This is still needed otherwise the following scenario will fail

Get-AzureRmEventGridDomain -ResourceGroupName aegBugbashResourceGroup2 -Name aegBugBashDomain2 | Get-AzureRmEventGridDomainTopic

File: src/ResourceManager/EventGrid/Commands.EventGrid/Domain/RemoveAzureEventGridDomain.cs (hunk 10 above)
Comment (ahamad-MS) [reply src/ResourceManager/EventGrid/Commands.EventGrid/Domain/RemoveAzureEventGridDomain.cs]: @cormacpayne  This is also needed otherwise the following scenario will break

Get-AzureRmEventGridDomain -ResourceGroupName aegBugbashResourceGroup2 -Name aegBugBashDomain2 | New-AzureRmEventGridDomainKey -KeyName Key1

File: src/ResourceManager/Sql/Commands.Sql/help/New-AzureRmSqlDatabase.md (hunk 11 above)
Comment (ahamad-MS) [reply src/ResourceManager/Sql/Commands.Sql/help/New-AzureRmSqlDatabase.md]: @cormacpayne  Fixed

**Comment by cormacpayne** [comment]
Sign job: https://azuresdkci.westus2.cloudapp.azure.com/job/ps-sign/1510/

**Comment by ahamad-MS** [comment]
@cormacpayne  All checks have passed including the sign job.
